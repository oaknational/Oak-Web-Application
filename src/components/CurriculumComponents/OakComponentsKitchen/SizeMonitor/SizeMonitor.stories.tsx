import { OakThemeProvider, oakDefaultTheme } from "@oaknational/oak-components";
import { Meta, StoryObj } from "@storybook/nextjs";
import { useState } from "react";

import SizeMonitor, { SizeMonitorBounds } from ".";

const meta: Meta<typeof SizeMonitor> = {
  component: SizeMonitor,
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof SizeMonitor>;

const TestComponent = () => {
  const [bbox, setBbox] = useState<SizeMonitorBounds | null>(null);

  return (
    <OakThemeProvider theme={oakDefaultTheme}>
      <SizeMonitor onChange={setBbox}>
        {bbox && (
          <div
            style={{
              height: "100vh",
              width: "100%",
              background: "red",
              padding: 12,
            }}
          >
            size={bbox.width}x{bbox.height}
          </div>
        )}
      </SizeMonitor>
    </OakThemeProvider>
  );
};

export const SimpleRadio: Story = {
  render: TestComponent,
};
