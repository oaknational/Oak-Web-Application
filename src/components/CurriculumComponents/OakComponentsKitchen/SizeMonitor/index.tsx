import { useEffect, useRef } from "react";

export type SizeMonitorBounds = Pick<DOMRect, "height" | "width">;

type SizeMonitorProps = {
  children: React.ReactNode;
  onChange: (bounds: SizeMonitorBounds) => void;
};
export default function SizeMonitor({
  onChange,
  children,
}: Readonly<SizeMonitorProps>) {
  const ref = useRef<HTMLDivElement>(null);
  useEffect(() => {
    if (ref.current) {
      const handler = () => {
        if (ref.current) {
          const boundingRect = ref.current.getBoundingClientRect();
          onChange({ width: boundingRect.width, height: boundingRect.height });
        }
      };
      const ro = new ResizeObserver(handler);
      ro.observe(ref.current);
      handler();
      return () => ro.disconnect();
    }
  }, [ref, onChange]);

  return (
    <div
      ref={ref}
      data-testid="size-container"
      style={{ position: "relative" }}
    >
      {children}
    </div>
  );
}
