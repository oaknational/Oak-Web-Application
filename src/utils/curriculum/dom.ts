export function createNode(
  nodeName: string,
  attributes: Record<string, string> = {},
) {
  const node = document.createElement(nodeName);
  for (const [attributeName, attributeValue] of Object.entries(attributes)) {
    node.setAttribute(attributeName, attributeValue);
  }
  return node;
}

export function anchorIntersectionObserver(callback: (id: string) => void) {
  const visibleAnchors = new Map();

  return (
    entries: Pick<
      IntersectionObserverEntry,
      "boundingClientRect" | "target" | "isIntersecting"
    >[],
  ) => {
    entries.forEach((entry) => {
      const top = entry.boundingClientRect.top;
      const year = entry.target.id;
      if (entry.isIntersecting) {
        visibleAnchors.set(year, top);
      } else {
        visibleAnchors.delete(year);
      }
      if (visibleAnchors.size > 0) {
        const sortedAnchors = Array.from(visibleAnchors.entries()).sort(
          ([, aTop], [, bTop]) => aTop - bTop,
        );
        callback(sortedAnchors[0]?.[0]);
      }
    });
  };
}

export function findContainingAnchor(root: Element) {
  let node: null | Element = root;
  while (node && node.tagName !== "A" && node.parentNode) {
    node = node.parentElement;
  }
  if (node instanceof HTMLAnchorElement) {
    return node;
  }
}

export function getAllTabFocusableElements(
  root?: HTMLElement | null,
): HTMLElement[] {
  if (root) {
    const selector =
      'a[href], button, input, textarea, select, details, [tabindex]:not([tabindex="-1"])';
    // <https://zellwk.com/blog/keyboard-focusable-elements/>
    return [...root.querySelectorAll(selector)] as HTMLElement[];
  }
  return [];
}
