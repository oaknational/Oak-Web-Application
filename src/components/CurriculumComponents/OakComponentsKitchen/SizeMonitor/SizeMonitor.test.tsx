import { waitFor } from "@testing-library/dom";

import SizeMonitor from ".";

import renderWithTheme from "@/__tests__/__helpers__/renderWithTheme";

const width = 100;
const height = 100;
const observeSpy = jest.fn();
window.ResizeObserver = class MockedResizeObserver {
  constructor(cb: ResizeObserverCallback) {
    setTimeout(() => {
      cb(
        [
          {
            contentRect: {
              height,
              width,
            },
          },
        ] as ResizeObserverEntry[],
        this,
      );
    }, 0);
  }
  // Attaching spy to "observe" function.
  observe = observeSpy;
  unobserve = jest.fn();
  disconnect = jest.fn();
};

describe("SizeMonitor", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  test("render", () => {
    const { baseElement } = renderWithTheme(
      <SizeMonitor onChange={() => {}}>testing</SizeMonitor>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test("resize", async () => {
    const spy = jest.spyOn(
      window.HTMLElement.prototype,
      "getBoundingClientRect",
    );
    spy.mockImplementation(
      () =>
        ({
          bottom: height,
          height,
          right: width,
          width,
          left: 0,
          top: 0,
        }) as DOMRect,
    );

    const onChange = jest.fn();
    const { getByTestId } = renderWithTheme(
      <SizeMonitor onChange={onChange}>
        <div style={{ width, height }}>testing</div>
      </SizeMonitor>,
    );

    expect(observeSpy).toHaveBeenCalledTimes(1);
    expect(observeSpy).toHaveBeenCalledWith(getByTestId("size-container"));

    await waitFor(() => {
      expect(onChange).toHaveBeenCalledWith({ width, height });
    });
  });
});
