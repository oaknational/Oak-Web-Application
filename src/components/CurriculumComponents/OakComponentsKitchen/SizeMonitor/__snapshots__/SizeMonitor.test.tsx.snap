// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SizeMonitor render 1`] = `
<body>
  <div>
    <div
      data-testid="size-container"
      style="position: relative;"
    >
      testing
    </div>
  </div>
</body>
`;
